migrate API and OAuth to use Service Load Balancers

joelddiaz opened the pull request:
Rather than use AWS-specific configuration for this pieces of the cluster, use the host cluster's native load balancer services.

This means the API and OAuth FQDNs will be the assigned names of the load balancers (ie no longer api.[clustername]....

Remove the AWS-specific DNS, load balancer, and target port install/uninstall code.

Commit:
move the api and oauth services to loadbalancer type

remove the AWS-specific DNS setup for the child cluster's API
remove the AWS-specific load balancer and target port configuration
remove the AWS-specific cleanup during uninstall (for the above)

change the OAuth URLs to use a dedicated URL (instead of piggybacking on top of the API URL)

Commit:
remove references to APINodeport

the kube-apiserver asset isn't used, and any consumer of this repo will have to bring their own kube-apiserver definition if they don't want ours
remove the kube-apiserver service asset
remove the param that would configure the kube-apiserver service

csrwng commented on pkg/aws/install.go:
```diff
@@ -364,15 +319,29 @@ func InstallCluster(name, releaseImage, dhParamsFile string, waitForReady bool)
 		return fmt.Errorf("cluster pod CIDR exceeds max address space")
 	}
 
+	apiDNSName, err := waitForServiceLoadBalancerDNS(client, name, kubeAPIServerServiceName)
```
s/waitForServiceLoadBalancerDNS/waitForServiceLoadBalancerAddress/
In GCP and Azure you'll get an IP and not a DNS name

csrwng commented on pkg/aws/wait.go:
```diff
@@ -159,3 +161,23 @@ func waitForClusterOperators(cfg *rest.Config) error {
+			return false, nil
+		}
+
+		serviceDNS = svc.Status.LoadBalancer.Ingress[0].Hostname
```
Return either `svc.Status.LoadBalancer.Ingress[0].Hostname` or `svc.Status.LoadBalancer.Ingress[0].IP` whichever is not blank. In Azure and GCP, you'll get the IP.

csrwng replied:
Also in the future we may want to rename the param `ExternalAPIAddress` instead of `ExternalAPIDNSName`

csrwng commented on pkg/aws/wait.go:
```diff
@@ -159,3 +161,23 @@ func waitForClusterOperators(cfg *rest.Config) error {
+func waitForServiceLoadBalancerDNS(client kubeclient.Interface, namespace, name string) (string, error) {
+	var serviceDNS string
+
+	err := wait.PollImmediate(10*time.Second, serviceLoadBalancerTimeout, func() (bool, error) {
```
It's preferable to use `clientwatch.UntilWithSync` as in https://github.com/openshift/hive/blob/5f17ab3a5fc96db6c670f81c105724213fcae4a1/contrib/cmd/waitforjob/main.go#L118
or 
https://github.com/openshift/hive/pull/953/files#diff-0a8cf6f710f69aa0899ff66b54ce3951R475

csrwng commented on pkg/aws/install.go:
```diff
@@ -175,7 +177,8 @@ func InstallCluster(name, releaseImage, dhParamsFile string, waitForReady bool)
 	log.Infof("Creating Kube API service")
 	apiNodePort, err := createKubeAPIServerService(client, name)
```
When the VPN udp->tcp change makes it in, we can also change the vpn load balancer and get rid of the need for the apiNodePort.

csrwng reviewed (commented):
This is very cool @joelddiaz ! Thx
Some comments

Commit:
address PR feedback

start handling the results that would come from a GCP/Azure load balancer (eg IP instead of Hostname)
stop refering to DNSNames for things, and just use Address (planning for GCP/Azure)
use more inteligent wait loop with listWatcher

Commit:
move VPN to service load blancer

remove aws-specific VPN configuration
remove aws-specific VPN teardown
move to Service Load Balancer for the VPN
remove the rendering of the vpn-service asset
remove the parameter to process the VPNNodePort (only used by VPN service asset)
remove udp handling when setting up listener

joelddiaz commented:
@csrwng thanks for all the feedback. I rebased with the VPN -> TCP changes and added the migration of VPN to type load balancer as well. Of course I addressed your PR feedback as well. We can squash once you've given it a look.

csrwng commented:
This is awesome! lgtm
Now just the router remains
